hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/qjournal/TestNNWithQJM.java

```diff
@@ -197,10 +197,9 @@ public void testMismatchedNNIsRejected() throws Exception {
           .manageNameDfsDirs(false).format(false).checkExitOnShutdown(false)
           .build();
       fail("New NN with different namespace should have been rejected");
-    } catch (ExitException ee) {
+    } catch (IOException ie) {
       GenericTestUtils.assertExceptionContains(
-          "Unable to start log segment 1: too few journals", ee);
-      assertTrue("Didn't terminate properly ", ExitUtil.terminateCalled());
+          "recoverUnfinalizedSegments failed for too many journals", ie);
```

Comment:
I wonder if we should modify the caller to catch the `IOException` and rethrow as `ExitException` to match previous behavior?

Reply:
I see you made this change by catching it in `FSNamesystem#loadFSImage()` but now I'm having second thoughts about it. The old behavior comes from this block in `FSEditLog#startLogSegment()`:
https://github.com/apache/hadoop/blob/63db1a85e376c2266afdc62b9590e40acc98429c/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java#L1437-L1447
Here we catch the `IOException` _within FSEditLog_ itself and add some useful contextual information, plus we clean up the `journalSet`. Seems useful. Given that `recoverUnclosedStreams()` is called from a variety of places, it seems that we can't just copy this approach by adding the `terminate(1)` logic within `FSEditLog#recoverUnclosedStreams()` -- e.g. `Checkpointer#doCheckpoint()` seems to assume it can catch an `IOException` from this phase and try again, and `NameNode#initializeSharedEdits()` / `NameNode#copyEditLogSegmentsToSharedDir()` have their own termination logic that we should probably respect. But it also doesn't really feel correct to me to add this special exception. What I am thinking now is to add a new flag to `recoverUnclosedStreams()`, like `boolean terminateOnFailure`, that we would set true in `FSNamesystem#startActiveServices()` and `FSImage#initEditLog()` and false for the three other callers I mentioned above. If `true`, we just rethrow the original `IOException`, but otherwise, we use similar logic as from `startLogSegment()` to add some additional context and call the termination. WDYT?